src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,629 @@
+        MessageFormat messageFormat = unpacker.getNextFormat();
+        ValueType valueType = messageFormat.getValueType();
+        int length;
+        Object item = null;
+         
+        switch(valueType) {
+            case NIL:
+                unpacker.unpackNil();
+                item = null;
+                break;
+            case BOOLEAN:
+                item = unpacker.unpackBoolean();
+                break;
+            case INTEGER:
+                switch (messageFormat) {
+                case UINT64:
+                    item = unpacker.unpackBigInteger();
+                    break;
+                case INT64:
+                case UINT32:
+                    item = unpacker.unpackLong();
+                    break;
+                default:
+                    item = unpacker.unpackInt();
+                    break;
+                }
+                break;
+            case FLOAT:
+                item = unpacker.unpackDouble();
+                break;
+            case STRING:
+                item = unpacker.unpackString();
+                break;
+            case BINARY:
+                length = unpacker.unpackBinaryHeader();
+                byte[] binaryValue = new byte[length];
+                unpacker.readPayload(binaryValue);
+                item = binaryValue;
+                break;
+            case ARRAY:
+                length = unpacker.unpackArrayHeader();
+                Object[] arrayValue = new Object[length];
+                for (int i = 0; i < length; i++) {
+                    arrayValue[i] = readValue(unpacker, new Object().getClass());
+                }
+                item = arrayValue;
+                //If the itemType is an array, we return an array. Else we convert the array to a list.
+                if (!itemType.isArray()) {
+                    item = new ArrayList<Object>(Arrays.asList(arrayValue));
+                }
+                break;
+            case MAP:
+                length = unpacker.unpackMapHeader();
+                Map<Object, Object> mapValue = new HashMap<Object, Object>();
+                for (int i = 0; i < length; i++) {
+                    Object key = readValue(unpacker, new Object().getClass());
+                    Object value = readValue(unpacker, new Object().getClass());
+                    mapValue.put(key, value);
+                }
+                item = mapValue;
+                break;
+            case EXTENSION:
+                /*
+                ExtensionTypeHeader extension = unpacker.unpackExtensionTypeHeader();
+                byte[] extensionValue = new byte[extension.getLength()];
+                unpacker.readPayload(extensionValue);
+                //Convert this to an object?
+                item = extensionValue;
+                */
+                throw new RuntimeException("Extension types are not supported yet");
+            default:
+                break;
+        }
+        if (itemType.isPrimitive()) {
```

Comment:
Add test(s) for all the primitive types, and see if we can remove the `toPrimitive` function :)

Reply:
Good call, we don't need this block (and in fact the way I'd been reading primitives was busted): https://github.com/dotnet/aspnetcore/pull/23532/commits/daa4199db56d8b3399940b0af00ab763e143c1c4